### PR TITLE
Update grammar of simple-const-expr according to 2020R1 specification

### LINKED ISSUE #13410: [Spec Deviation] Allow any simple-const-expr in describing singleton types
**Description:**
As per the spec (v0.990:2019-01-16 - page 29, 30), 

"singleton-type-descriptor := simple-const-expr

A singleton type is a type containing a single shape. A singleton type is described using an compile-time constant expression for a single value: the type contains the shape of that value. Note that it is possible for the variable-reference within the simple-const-expr to reference a structured value."

```
simple-const-expr :=
  nil-literal
  | boolean
  | [Sign] int-literal
  | [floating-point-type] [Sign] floating-point-literal
  | string-literal
  | variable-reference
floating-point-type := < (decimal|float) >
```

Unsupported expressions need to be supported.

**Affected Versions:**
0.990.2

### PR DESCRIPTION
**Pull request by chiranSachintha**
nil-literal
  | boolean-literal
  | [Sign] int-literal
  | [Sign] floating-point-literal
  | string-literal
  | constant-reference-expr

Fixes #13410 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/finite/FiniteTypeTest.java [hunk 1]
```diff
@@ -367,4 +367,21 @@ public void testEscapedTypeName() {
...
+    }
+    @Test()
+    public void testFiniteTypesWithPositiveFloats() {
+        BValue[] returns = BRunUtil.invoke(result, "testFiniteTypesWithPositiveFloats");
+        Assert.assertEquals(returns.length, 1);
+        Assert.assertNotNull(returns[0]);
+        Assert.assertTrue(returns[0] instanceof BFloat);
+        Assert.assertEquals(((BFloat) returns[0]).floatValue(), 1.5);
+    }
+
```
Comment (dulvinw) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/finite/FiniteTypeTest.java:386]: We usually try to do the testing on ballerina itself. And panic if not the intended result of the test.

Comment (dulajdilshan) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/finite/FiniteTypeTest.java:386]: +1

File: tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/finite/finite-type.bal [hunk 2]
```diff
@@ -461,6 +461,31 @@ function testFiniteTypesWithDiscriminatedMembers() returns [any, any, any, any,
...
+type d +3|+5;
+
+function testFiniteTypesWithPositiveIntegers() {
+    d n = +3;
+    d comparator = +3;
+    if (n == comparator) {
+       n = +5;
+    }
+    if (n != 5){
+        panic error("Type mismatch");
```
Comment (MaryamZi) [inline tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/finite/finite-type.bal:473]: Best practices for errors
- as the error reason use meaningful simple strings with no spaces
- as the message field give additional details about what caused the error

See https://github.com/ballerina-platform/ballerina-lang/blob/7fa2ee515984b42c1281a43621753838474f6907/tests/jballerina-bstring-unit-test/src/test/resources/test-src/expressions/listconstructor/list_constructor_infer_type.bal#L64 for an example.

File: tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/finite/finite-type.bal [hunk 3]
```diff
@@ -461,6 +461,31 @@ function testFiniteTypesWithDiscriminatedMembers() returns [any, any, any, any,
...
+
+type f +1.2|+1.5;
+
+function testFiniteTypesWithPositiveFloats() {
+    f n = +1.2;
+    f comparator = +1.2;
+    if (n == comparator) {
+       n = +1.5;
+    }
+    if (n != 1.5){
```
Comment (MaryamZi) [inline tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/finite/finite-type.bal:485]:
```suggestion
    if (n != 1.5) {
```

File: tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/finite/finite-type.bal (hunk 2 above)
Comment (chiranSachintha) [reply tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/finite/finite-type.bal]: Fixed

File: tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/finite/finite-type.bal [hunk 4]
```diff
@@ -461,6 +461,31 @@ function testFiniteTypesWithDiscriminatedMembers() returns [any, any, any, any,
     return [a, b, c, d, e];
 }
 
+type d +3|+5;
```
[the remark this answers is not in the document]
Comment (chiranSachintha) [reply tests/jballerina-bstring-unit-test/src/test/resources/test-src/types/finite/finite-type.bal]: Fixed

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/finite/FiniteTypeTest.java (hunk 1 above)
Comment (chiranSachintha) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/types/finite/FiniteTypeTest.java]: Done

**Pull request merged**